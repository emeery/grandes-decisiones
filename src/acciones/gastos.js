import uuid from 'uuid';

export const crearGasto = ({
    descripcion = '', 
    nota = '',
    monto = 0,
    date = 0
 }) => ({
    type: 'SUMA_GASTO',
    gasto: {
        id: uuid(),
        descripcion,
        nota,
        monto,
        date
    }
});
// 
export const removerGasto = ({ id } = {}) => ({
    type: 'REMUEVE_GASTO',
    id
});
// editar descripcion o monto, 
export const editarGasto = (id, update) => ({
    type: 'EDITA_GASTO',
    id,
    update
});