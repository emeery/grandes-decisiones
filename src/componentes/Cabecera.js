import React from 'react';
import {NavLink} from 'react-router-dom';
<<<<<<< HEAD
// una nueva nota pendiente, el oscuro fondo tomaba un aspecto lúgubre 
// y la unica fuente de luz que despedía 3#sx1$. dotted edge 

const Cabecera = () => (
    <header>
    <h1>Portafolio</h1>
    <NavLink to='/' activeClassName='is-active' exact={true}>Hogar</NavLink>
    <NavLink to='/portafolio' activeClassName='is-active' exact={true}>Portafolio</NavLink>
    <NavLink to='/contacto'activeClassName='is-active'>Contacto</NavLink>
=======

// <NavLink to="/ayuda" activeClassName='is-active'>Ayuda</NavLink>
const Cabecera = () => (
    <header>
    <h1>Cuales son tus gastos ?</h1>
    <h3>estrategias para el ahorro</h3>
    <NavLink to="/" activeClassName='is-active' exact={true} >Escritorio</NavLink>
    <NavLink to="/crealo" activeClassName='is-active'>Crear Gastos</NavLink>
    
>>>>>>> 0608cb1624129c0297eabba445c5e9bc3f131dc5
    </header>

); export default Cabecera;

