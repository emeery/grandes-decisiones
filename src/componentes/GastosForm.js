import React from 'react';
import moment from 'moment';
import { SingleDatePicker} from 'react-dates'; 

// const ahora = moment();
// console.log(ahora.format('MMM Do, YYYY') );

class GastosForm extends React.Component {
    constructor(props) {
        super( props );
        this.state = {
        descripcion: props.gasto ?  props.gasto.descripcion : '' , 
        nota: props.gasto ? props.gasto.nota : '' , 
        monto: props.gasto ? (props.gasto.monto / 100 ).toString() : '', 
        date: props.gasto ? moment(props.gasto.date) : moment(),
        calendarFocused : false,
        error: ''
        };
    }
    descripcionCambio = (e) => {
        const descripcion = e.target.value; 
        this.setState(() => ({ descripcion })); 
    };
    // persistncia ();
    notaCambio = (e) => {
        const nota = e.target.value;
        this.setState(()=> ({ nota }));
    };
    montoCambio = (e) => {
        const monto = e.target.value;
        if (!monto || monto.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ monto }));
        }
    };
    onDatesChange = (date) => {
        if(date) {
            this.setState(()=> ({ date }));
        }
    };
    // focused - requerido
    onFocusChange = ({focused}) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.descripcion || !this.state.monto) {
            this.setState(() => ({error: 'ingresa los 2 valores'}));
        } else {
            this.setState(() => ({error: ''}));
            this.props.onSubmit({
                descripcion: this.state.descripcion,
                monto: parseFloat(this.state.monto, 10) * 100,
                date: this.state.date.valueOf(), // nmero regular que representa esa fecha
                nota: this.state.nota
            });
        }
    };
    render() {
        return( <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>      
            <input 
            type='text'
            placeholder='Descripcion'
            autoFocus
            value={this.state.descripcion}
            onChange={this.descripcionCambio} 
            />
            <input
            type='text'
            placeholder='Monto'
            value={this.state.monto}
            onChange={this.montoCambio}
            />
            
            <SingleDatePicker
            date={this.state.date}
            onDateChange={this.onDatesChange}
            focused={this.state.calendarFocused} 
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={()=>false}
            />
            
            <textarea
            placeholder='agrega una nota para tu gasto'
            value={this.state.nota}
            onChange={this.notaCambio}
            >
            </textarea>

            <button>Agrega Gasto</button>
        </form>    
        </div>);
    }
}; export default GastosForm;


// crear playlist.
// nombrePLaylist /* -string, nombre 1/100*/
// <button /*  elegir imagen */
// descripcion /* -string estos tres 0/300 */
// <button /* cancelar */
// <button /* crear */
