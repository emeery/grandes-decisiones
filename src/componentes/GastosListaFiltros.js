import React from 'react';
import {connect} from 'react-redux';
import {DateRangePicker} from 'react-dates'; 

import {
    establecerTextoFiltros, 
    ordenarPorMonto, 
    ordenarPorFecha,
    establecerFechaInicio, 
    establecerFechaFin
} from '../acciones/filtros';

export class GastosListaFiltros extends React.Component {
    state = {
        calendarFocused: null
    };
    // this.props.dispatch
    onDatesChange = ({startDate, endDate}) => {
        this.props.establecerFechaInicio(startDate);
        this.props.establecerFechaFin(endDate);
    };
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    };
    onTextoChange = (e) => {
        this.props.establecerTextoFiltros(e.target.value);
    };
    onSortChange = (e) => {
        if (e.target.value === 'date') {
            this.props.ordenarPorFecha();
        } else if (e.target.value === 'monto') {
            this.props.ordenarPorMonto();
        }
    };
    render() {
        return  ( 
        <div>
        
        <input 
        type='text' 
        value={this.props.filtros.texto}
        onChange={this.onTextoChange}
        />
        
        <select 
        value={this.props.filtros.ordenarPor}
        onChange={this.onSortChange}
        >
        
        <option value='date'>
            fecha </option>
        <option value='monto'>
            monto </option>
        
        </select>
        <DateRangePicker
        startDate={this.props.filtros.startDate}
        endDate={this.props.filtros.endDate}
        onDatesChange={this.onDatesChange}
        focusedInput={this.state.calendarFocused}
        onFocusChange={this.onFocusChange}
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
        />
    </div> );
    }
};

const mapStateToProps = (state) => ({
    filtros: state.filtros });
const mapDispatchToProps = (dispatch) => ({
    establecerTextoFiltros: (texto) => dispatch(establecerTextoFiltros(texto) ),
    ordenarPorMonto: () => dispatch(ordenarPorMonto() ),
    ordenarPorFecha: () => dispatch(ordenarPorFecha() ),
    establecerFechaInicio: (startDate) => dispatch(establecerFechaInicio(startDate) ),
    establecerFechaFin: (endDate) => dispatch(establecerFechaFin(endDate) )
});
export default connect(mapStateToProps, mapDispatchToProps)(GastosListaFiltros);