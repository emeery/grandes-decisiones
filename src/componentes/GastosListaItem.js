import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
const GastosListaItem = ({
    id, descripcion, monto, date}) => (
    <div>
        <Link to={`/edita/${id}`}>
        <h3> {descripcion} </h3>
        </Link>
        <p> 
            {numeral(monto / 100 ).format('$0,0.00')} 
        - 
            {moment(date).format('MMMM Do, YYYY')}
        </p>
       
    </div>
);
export default GastosListaItem;

// const GastosListaItem = ({ 
//      dispatch, id, descripcion, monto, fecha}) => (
//      <div>
//         <h3> descripcion </h3>
//         <h2>fecha: {fecha} - monto: {monto}</h2>
//         <button onClick={() => {
//             dispatch(removeGasto({ id }));
//         }}
//         > remover </button>
//     </div>
// );
// export default connect()(GastosListaItem);

// <h2> monto fecha </h2>
// <button onClick={() => {
//     dispatch(removerGasto({ id }));
// }}
// > remover </button>