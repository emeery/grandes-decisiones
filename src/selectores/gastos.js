import moment from 'moment';
// valida las fechas y el texto /
// seleccionaGasto
export default (gastos, {texto, ordenarPor, startDate, endDate}) => {
    return gastos.filter((gasto) => {
        const creadoEnElMomento = moment(gasto.date);
        const fechaInicioMatch = startDate ? startDate.isSameOrBefore(creadoEnElMomento, 'day') : true;
        const fechaFinMatch = endDate ? endDate.isSameOrAfter(creadoEnElMomento, 'day'): true;
        const textoMatch = gasto.descripcion.toLowerCase().includes(texto);
        return textoMatch && fechaInicioMatch && fechaFinMatch;
    }).sort(( a , b )=> {
        if (ordenarPor === 'date') {
            return a.date < b.date ? 1 : -1;
        } else if (ordenarPor === 'monto') {
            return a.monto < b.monto ? 1 : -1;
        }
    });
};

// The selectors are functions that take the status of Redux 
// as an argument and return some data to pass to the component.

// const fechaInicioMatch = typeof fechaInicio !== 'number' || gasto.fecha >= fechaInicio;
// const fechaFinMatch = typeof fechaFin !== 'number' || gasto.fecha <= fechaFinMatch;
