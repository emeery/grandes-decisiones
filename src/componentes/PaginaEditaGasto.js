import React from 'react';
import {connect} from 'react-redux';
import GastosForm from './GastosForm';
import {
    editarGasto, 
    removerGasto 
} from '../acciones/gastos';

export class PaginaEditaGasto extends React.Component {
    onSubmit = ( gasto ) => {
        this.props.editarGasto(this.props.gasto.id, gasto);
        this.props.history.push('/');
    };
    removerEsto = () => {
        this.props.removerGasto({id: this.props.gasto.id});
        this.props.history.push('/');
    };
    render() {
        return(<div>
            <GastosForm
            gasto={this.props.gasto}
            onSubmit={this.onSubmit}
            />
            <button 
            onClick={this.removerEsto} 
            > X </button>
        </div>
        );
    }
};

// subconjuntos
const mapStateToProps = (state, props) => ({
    gasto: state.gastos.find((gasto) => gasto.id === props.match.params.id) 
});

const mapDispatchToProps = (dispatch, props) => ({
    editarGasto: (id, gasto) => dispatch(editarGasto(id, gasto)),
    removerGasto: (data) => dispatch(removerGasto( data ))
});

export default connect(mapStateToProps, mapDispatchToProps )(PaginaEditaGasto);


