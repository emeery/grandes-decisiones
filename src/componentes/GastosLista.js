import React from 'react';
import { connect } from 'react-redux';
import GastosListaItem from './GastosListaItem';
import seleccionaGastos from '../selectores/gastos';

// conecta tu componente a la tienda redux
const GastosLista = (props) => (
    <div><h1>Expendio Lista</h1>
        {props.gastos.length}
        {props.gastos.map((gastitos) => {
           return <GastosListaItem 
                key={gastitos.id} {...gastitos}/>
        })}
    </div>
); 
// definimos lo que queremos de la tienda
// y definimos el componnte GastosLista 

const mapStatetoProps = (state) => {
    return { 
        gastos: seleccionaGastos(state.gastos, state.filtros)
    };
};
export default connect(mapStatetoProps)(GastosLista);
