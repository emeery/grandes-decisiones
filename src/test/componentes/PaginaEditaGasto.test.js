import React from 'react';
import {shallow} from 'enzyme';
import gastos from '../accesorios/gastos';
import {PaginaEditaGasto} from '../../componentes/PaginaEditaGasto';
//import { removerGasto } from '../../acciones/gastos';
// paises gerardo
// ghana, egipto, 
let editarGasto, remueveGasto, envoltura, history;
// spies
beforeEach(() => {
    editarGasto = jest.fn();
    remueveGasto = jest.fn();
    history = { push: jest.fn() };
    envoltura = shallow(
        <PaginaEditaGasto 
        editarGasto={editarGasto} 
        removerGasto={remueveGasto}
        history={history}
        gasto={gastos[2]}
        />
    );
});
test('snapshot PaginaEditaGasto', ()=>{
    expect(envoltura).toMatchSnapshot();
});

// this.props.editarGasto(this.props.gasto.id, gasto);
test('deberia controlar editarGasto', ()=>{
    envoltura.find('GastosForm').prop('onSubmit')(gastos[2]);
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editarGasto).toHaveBeenLastCalledWith(gastos[2].id, gastos[2]);
    //oltura).toMatchSnapshot();
});

test('deberia controlar removerGasto', ()=>{
    envoltura.find('button').simulate('click');
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(remueveGasto).toHaveBeenLastCalledWith({
        id: gastos[2].id
    });
});