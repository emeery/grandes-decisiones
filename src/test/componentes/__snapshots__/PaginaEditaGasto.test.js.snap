// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot PaginaEditaGasto 1`] = `
<div>
  <GastosForm
    gasto={
      Object {
        "date": 345600000,
        "descripcion": "tarjeta credito",
        "id": "3",
        "monto": 4500,
        "nota": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
     X 
  </button>
</div>
`;
